### PR TITLE
[Profiler] Enable profiler with fine grained profile

### LINKED ISSUE #1573: profiler needs maintenance
profiler is not working due to massive changes in the core.
we should have `-Denable_profile=true` build somewhere in ci to check to maintain profiler

### PR DESCRIPTION
**Pull request by zhoonit**
- [Profiler] Enable profiler with fine grained profile

```
This patch enables profiler with layer profiling.

**Self evaluation:**
1. Build test: [X]Passed [ ]Failed [ ]Skipped
2. Run test: [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: Jihoon Lee <jhoon.it.lee@samsung.com>

```

resolves #1573

### CODE REVIEW

File: Applications/MNIST/jni/main.cpp [hunk 1]
```diff
@@ -291,6 +297,7 @@ int main(int argc, char *argv[]) {
     std::cerr << "Error during train " << e.what() << std::endl;
     return 0;
   }
+  std::cout << listener;
```
Comment (kparichay) [inline Applications/MNIST/jni/main.cpp:300]: wrap with #ifdef

File: nntrainer/layers/layer_node.cpp [hunk 2]
```diff
@@ -18,13 +18,17 @@
...
 #include <node_exporter.h>
+#include <profiler.h>
 #include <time_dist.h>
 #include <util_func.h>
 
 #include <base_properties.h>
 #include <common_properties.h>
 
 namespace nntrainer {
+static constexpr const char *FORWARD_SUFFIX = ":forward";
```
Comment (kparichay) [inline nntrainer/layers/layer_node.cpp:29]: how about using function name instead of predefined strings?

**Review by kparichay — APPROVED** [review]
LGTM

File: nntrainer/layers/layer_node.cpp (hunk 2 above)
Comment (zhoonit) [reply nntrainer/layers/layer_node.cpp:29]: you mean __func__? registering point doesn't have access to actual func name (for performance concern) where it is executed, so we will need a predefined strings.

File: Applications/MNIST/jni/main.cpp (hunk 1 above)
Comment (zhoonit) [reply Applications/MNIST/jni/main.cpp:300]: Right, updated.

File: nntrainer/layers/layer_node.cpp (hunk 2 above)
Comment (kparichay) [reply nntrainer/layers/layer_node.cpp:29]: alright.

File: nntrainer/layers/layer_node.cpp [hunk 3]
```diff
@@ -439,6 +443,17 @@ InitLayerContext LayerNode::finalize(const std::vector<TensorDim> &input_dims) {
...
 
   layer->finalize(init_context);
+
+#ifdef PROFILE
+  auto profile_name = [this](const char *suffix) {
+    return getName() + suffix + "(" + getType() + ")";
+  };
+#endif
+
+  REGISTER_EVENT(profile_name(FORWARD_SUFFIX), forward_event_key);
```
[the remark this answers is not in the document]
Comment (zhoonit) [reply nntrainer/layers/layer_node.cpp:453]: header (profiler.h) nullifies this so it's okay

[the remark this answers is not in the document]
Comment (lhs8928) [reply nntrainer/layers/layer_node.cpp:453]: I mean not because of REGISTER_EVENT but for profile_name

[the remark this answers is not in the document]
Comment (zhoonit) [reply nntrainer/layers/layer_node.cpp:453]: yes, even forward_event_key is defined only for #ifdef PROFILE.

**Review by lhs8928 — APPROVED** [review]
LGTM

**Pull request merged**